### DISCUSSION TITLE
Chain command after optional "--"

### DISCUSSION THREAD

**Discussion by elieux** [chrisant996/clink#334] [Q&A] [answered]
Hey.  I'm trying to write completions for *aws-vault*, but the question is probably applicable to other software as well.  What is the best way to match either of:

1. a chained command (`aws-vault [<flags>] exec [<flags>] foo aws s3 ls`)
2. a "--" followed by a chained command (`aws-vault [<flags>] exec [<flags>] foo -- aws s3 ls`)

The skeleton of the script representing the question is pasted below (the whole script is at https://gist.github.com/elieux/cb864345878f501ff8fe1ab728d52e01).  I tried various modifications unsuccessfully and this is the only one that worked at least somewhat.  Problems are: it covers only the second case and it doesn't suggest using "--".

```lua
local function profile()
    return { "foo", "bar" } -- this is actually found out dynamically
end

clink.argmatcher("aws-vault")
:addflags("--foo")
:addarg({
    "exec" .. clink.argmatcher()
    :addflags("--bar")
    :addarg(profile)
    :addflags(
        "--" .. clink.argmatcher():chaincommand()
    )
    :nofiles(),
})
:nofiles()
```

**Comment by chrisant996** [comment]
I'm having trouble following the question, because I don't know anything about `aws-vault` or its syntax.

I think you mean that `aws-vault exec foo` tells aws-vault that everything after that (e.g. `aws s3 ls`) is interpreted as a separate command line, and so the `aws` should look for an argmatcher and do completions accordingly past the `aws` part.

And I think you mean that `exec` has one argument slot that must be provided, and then starting at argument slot 2 it should actually chain to a different argmatcher.

On the argmatcher that's linked to `"exec"`, change the `:nofiles()` to `:chaincommand()`.

**Reply by elieux** [reply]
Now *I'm* not sure whether I understood.  I changed the script, but I'm getting the same results:

1. With `aws-vault exec foo <TAB>`, I'm not getting anything and `aws-vault exec foo cmd` is not recognizing `cmd` as a command.
2. With `aws-vault exec foo -- <TAB>`, I'm not getting anything (but that's a different issue since it doesn't really work with `cmd /c <TAB>` either). With `aws-vault exec foo -- cmd`, `cmd` is recognized as a command.

```lua
local function profile()
    return { "foo", "bar" } -- this is actually found out dynamically
end

clink.argmatcher("aws-vault")
:addflags("--foo")
:addarg({
    "exec" .. clink.argmatcher()
    :addflags("--bar")
    :addarg(profile)
    :addflags(
        "--" .. clink.argmatcher():chaincommand()
    )
    :chaincommand(),
})
:nofiles()
```

**Comment by chrisant996** [comment]
And regarding not suggesting `--`, I think that you only would want `--` to be suggested if you couldn't make it chain automatically after the first arg slot.  So the preceding suggestion should resolve both of the problems, yes?

**Reply by elieux** [reply]
I actually prefer writing "--", but since both are possible, I'd like to know if there's a way to have both.

**Comment by chrisant996** [comment, marked as answer]
Oops, my mistake:

I said to replace `:nofiles()` with `:chaincommand()`.  But there is a bug that in some cases falls back to the "parent" argmatcher by mistake if `:nofiles()` isn't used as well.

For now you can use `:chaincommand():nofiles()`.  And the next Clink update will include a fix so that the `:nofiles()` isn't needed (it should be implied by `:chaincommand()`).

But also, `aws-vault exec foo `<kbd>Tab</kbd> will behave as though `exec.enable` is False, due to a bug that is already fixed for the upcoming v1.3.40 release.  It chains correctly, it just didn't respect `exec.enable` correctly after chaining.

This script does what I think are you describing:
```lua
local function profile()
    return { "foo", "bar" } -- this is actually found out dynamically
end

local exec_parser = clink.argmatcher()
    :addflags(
        "--bar",
        "--" .. clink.argmatcher():chaincommand()
    )
    :addarg(profile)
    :chaincommand()

clink.argmatcher("aws-vault")
:addflags("--foo")
:addarg({"exec" .. exec_parser})
:nofiles()
```

**Note:** I rearranged things a little, just for clarity.  For example, `:addflags("--bar")` then `:addarg(profile)` then `:addflags("--")` is the same as `:addflags("--bar", "--"):addarg(profile)`.

Did you want for `--` to only be a possible completion after a profile argument?  If so, then I don't know a way to easily allow certain flags only at certain places in an argmatcher using the current Clink APIs.

It's _possible_, but it's not easy.  Currently you'd have to use make the `profile()` function return a table where each match is linked to an argmatcher that redefines the flags available (I'll call it the "dashdash argmatcher").  And since it's found out dynamically, you'd need to use a "delayinit" argmatcher, and it would need to `:reset()` itself at the right times to be able to re-generate a new table of profiles and link them with the dashdash argmatcher.  Probably more effort than it's worth.

**Reply by elieux** [reply]
> For example, `:addflags("--bar")` then `:addarg(profile)` then `:addflags("--")` is the same as `:addflags("--bar", "--"):addarg(profile)`.

Yeah, I couldn't find how these stack in the docs.

Same with how multiple `:addarg()` stack even though it's in the very first example. I hope I'm not being a dummy here, but I think the docs would benefit if the whole set of command lines possible with that argmatcher is shown (or at least if the stacking mechanisms are explained in two more sentences). Also, where does `wonder` come from there?

> `Did you want for -- to only be a possible completion after a profile argument?`

Yes, that's how `exec` works, the profile name is a mandatory argument before the chained command. Would it be possible to use `--` as an arg instead of a flag?

> It's *possible*, but it's not easy.

I might've found an easy way actually. It didn't want to work like this with "function as argument option", but `delayinit` made it work. In the screenshot, `sww` is an actual profile.

![image](https://user-images.githubusercontent.com/1399406/185580027-ec367416-8077-469e-92f7-cab68a3ff22e.png)

```lua
local map = require("funclib").map

local function profile()
    return { "foo", "bar" } -- this is actually found out dynamically
end

clink.argmatcher("aws-vault")
:addflags("--foo")
:addarg({
    "exec" .. clink.argmatcher()
    :addflags("--bar")
    :addarg({ delayinit=function()
        return map(profile(), function(item)
            return item .. clink.argmatcher()
            :addflags(
                "--" .. clink.argmatcher():chaincommand():nofiles()
            ):chaincommand():nofiles()
        end)
    end })
    :nofiles(),
})
:nofiles()
```

**Reply by elieux** [reply]
> I said to replace `:nofiles() with :chaincommand()`. But there is a bug that in some cases falls back to the "parent" argmatcher by mistake if `:nofiles()` isn't used as well.
> 
> For now you can use `:chaincommand():nofiles()`. And the next Clink update will include a fix so that the `:nofiles()` isn't needed (it should be implied by `:chaincommand()`).

Regarding this, I just noticed that `:nofiles()` does something different that the name implies. Docs mention it "dead-ends" a matcher, but I didn't expect it to also prevent the parent matcher from continuing. In the example snippet below, after using the `--region=...` flag, nothing is matched anymore. Is this something worth explaining further in the docs?

```lua
:addflags(
    "--region=" .. clink.argmatcher():addarg(region):nofiles(),
)
:addarg(profile)
:nofiles()
```

**Reply by chrisant996** [reply]
> > For example, `:addflags("--bar")` then `:addarg(profile)` then `:addflags("--")` is the same as `:addflags("--bar", "--"):addarg(profile)`.
> 
> Yeah, I couldn't find how these stack in the docs.
> 
> Same with how multiple `:addarg()` stack even though it's in the very first example. I hope I'm not being a dummy here, but I think the docs would benefit if the whole set of command lines possible with that argmatcher is shown (or at least if the stacking mechanisms are explained in two more sentences). Also, where does `wonder` come from there?

The section [The Basics](https://chrisant996.github.io/clink/clink.html#argmatcher_basics) mentions that arguments are positional.

I can try to come up with phrasing that makes it clear that arguments are positional, and therefore the order of `:addarg()` matters, and that flags are _not_ positional and therefore order has no meaning for flags.

> > It's _possible_, but it's not easy.
> 
> I might've found an easy way actually. It didn't want to work like this with "function as argument option", but `delayinit` made it work. In the screenshot, `sww` is an actual profile.

You did what I suggested.  (I suggested `delayinit`, not function as argument option.)

But, you haven't added a `:reset()` yet, so completions are only generated one time, linked with a chaincommand argmatcher, and then all future `aws-vault` commands will use that same set of completions.  If the profile completions are meant to be generated on the fly each time, then that is not happening anymore.  Function as argument option cannot return argmatchers (nor matches linked with argmatchers), so `:reset()` is needed to clear the argmatcher, and then you can build a new argmatcher with newly generated profile completions.

You can use `:setdelayinit()` to give the overall argmatcher a chance to `:reset()` if it wants to, and then reconstruct itself and use `delayinit` in the `:addarg()` so that it doesn't generate profile matches until/unless completions are needed for that argument position.

Also, there is an opportunity to speed it up:
> ```lua
> ...
>                 "--" .. clink.argmatcher():chaincommand():nofiles()
> ...
> ```
The line above is happening inside a loop.  So, it ends up creating two whole new argmatchers for each match.  You can speed it up by creating the argmatchers outside the loop, and use them over and over inside the loop:  (`:map()` loops over all of the matches)

```lua
local map = require("funclib").map

local function profile()
    return { "foo", "bar" } -- this is actually found out dynamically
end

clink.argmatcher("aws-vault")
:addflags("--foo")
:addarg({
    "exec" .. clink.argmatcher()
    :addflags("--bar")
    :addarg({ delayinit=function()
        -- Build argmatcher outside the map() loop.
        local chain = clink.argmatcher()
                :addarg("--" .. clink.argmatcher():chaincommand():nofiles())
                :chaincommand():nofiles()
        return map(profile(), function(item)
            return item .. chain -- Reuse existing argmatcher inside the map() loop.
        end)
    end })
    :nofiles(),
})
:nofiles()
```

**Reply by elieux** [reply]
> The section [The Basics](https://chrisant996.github.io/clink/clink.html#argmatcher_basics) mentions that arguments are positional.

> 
> I can try to come up with phrasing that makes it clear that arguments are positional, and therefore the order of `:addarg()` matters, and that flags are not positional and therefore order has no meaning for flags.

That was kind of what I meant. I read "positional" and "position independent" and all, but it didn't immediately give me the knowledge of:

- `:addarg()` specifies an argument and the available options for it (especially with the `wonder` part there)
- multiple `:addarg()` on a matcher add separate arguments in that order
- multiple `:addflags()` on a matcher will merge into a single set

This really wasn't obvious to me even after having read that part of the docs multiple times. Now when I know, I kind of see it there, but I recommend really giving the *basics* to the reader as well.

> You did what I suggested. [...] But, you haven't added a `:reset()` yet

Right. I'll try to play with resetting later, but right now, this thing is very much Good Enough™ for me.

> Function as argument option cannot return argmatchers (nor matches linked with argmatchers)

I would expect this to be explicitly mentioned in the docs on *Functions As Argument Options*. I can see how "table of potential matches" could maybe do that for someone, but given that this section directly follows two sections on linking matchers and the vocabulary isn't established, explicit sounds better to me in this case.

> Also, there is an opportunity to speed it up

Good point.


#### In conclusion

Would it be possible to somehow use `--` as an arg instead of a flag, or is the solution presented here the canonical one for this case?

I'll be looking forward to v1.3.40 to see those two issues fixed and I'll revisit my script with those two suggestions in mind.

I'd like to use my first-time experience to improve the docs. I'm willing to help out (give drafts, more suggestions, revisions...) if you're open to that.

**Reply by chrisant996** [reply]
> I think the docs would benefit if the whole set of command lines possible with that argmatcher is shown (or at least if the stacking mechanisms are explained in two more sentences).

"the whole set of command lines possible" -- The whole set is very large, and it isn't fair to a reader to dump all those subtly different cases on them so early. Descriptions seem like a better approach.

With that said, the first example specifically calls out "arg # 1" and "arg # 2".

I'm worried that part of the complication may come from many readers trying to learn Clink's API and also the Lua language at the same time.  That's very difficult for the docs to overcome with just a couple of sentences.

> Also, where does `wonder` come from there?

Oops, it's a typo from an earlier attempt at clarifying the example.

> Regarding this, I just noticed that `:nofiles()` does something different that the name implies. Docs mention it "dead-ends" a matcher, but I didn't expect it to also prevent the parent matcher from continuing. In the example snippet below, after using the `--region=...` flag, nothing is matched anymore. Is this something worth explaining further in the docs?

"Dead end" and "suggest no completions" means parsing ends.  This feels like English ambiguity:  it could be interpreted as "this parser doesn't provide completions, but other parsers still can", or it could be interpreted as "the parser provides completions, but the set of completions is empty and no more parsing occurs".  The latter is the intended meaning.

I'll try to rephrase, but I don't want to turn it into a paragraph of details.

> > Function as argument option cannot return argmatchers (nor matches linked with argmatchers)
>
> I would expect this to be explicitly mentioned in the docs on Functions As Argument Options. I can see how "table of potential matches" could maybe do that for someone, but given that this section directly follows two sections on linking matchers and the vocabulary isn't established, explicit sounds better to me in this case.

I think there's an easier and more concise way to clarify it, and I'll do that.

> Would it be possible to somehow use `--` as an arg instead of a flag, or is the solution presented here the canonical one for this case?

It's possible to add `--` as an arg.  But if you also add flags that start with `-` then it will be confusing.  Clink only completes flags when you type a character that the defined flags start with, and then it only completes flags.  So your suggestion could lead to `--` being omitted from the list of possible completions if you type `-` and there are any other flags defined.

But what problem is intended to be solved by calling `--` an argument instead of a flag?  Maybe there are other ways to solve the problem?

**Comment by elieux** [comment]
> It's possible to add `--` as an arg.

You're right, it's possible, but not in an obvious way.  If I naively match a profile arg, then the `--` arg and then chain a command, I just get an infinite series of `--` in the completion.  This doesn't happen if I change `--` into a word like `aa`.

```lua
local function profile()
    return { "foo", "bar" } -- this is actually found out dynamically
end

clink.argmatcher("aws-vault")
:addflags("--foo")
:addarg({
    "exec" .. clink.argmatcher()
    :addflags("--bar")
    :addarg(profile)
    :addarg({ "--" }) -- same result if I link clink.argmatcher():chaincommand():nofiles() here
    :chaincommand():nofiles(),
})
:nofiles()
```

It works (completes `--` just once and then recognizes the chained command) if I link the `--` arg after each profile arg option as in our current solution.

```lua
local map = require("funclib").map

local function profile()
    return { "foo", "bar" } -- this is actually found out dynamically
end

clink.argmatcher("aws-vault")
:addflags("--foo")
:addarg({
    "exec" .. clink.argmatcher()
    :addflags("--bar")
    :addarg({ delayinit=function()
        return map(profile(), function(item)
            return item .. clink.argmatcher()
            :addarg({ "--" }) -- same result if I link clink.argmatcher():chaincommand():nofiles() here
            :chaincommand():nofiles()
        end)
    end })
    :nofiles(),
})
:nofiles()
```

Given my current understanding of args, I don't see why the first example fails.  The second example is more complicated and also cannot continue with completion without exactly matching an existing profile.

> But what problem is intended to be solved by calling `--` an argument instead of a flag? Maybe there are other ways to solve the problem?

I was hoping it would:

1. simplify the matcher, see above
2. allow autocompleting `--` arg right away (not only after writing the first `-`)

There's also the question of how to make `--` optional in this case, there doesn't seem to be a dedicated API for this.

**Reply by chrisant996** [reply]
> If I naively match a profile arg, then the `--` arg and then chain a command, I just get an infinite series of `--` in the completion. This doesn't happen if I change `--` into a word like `aa`.

It's because the example adds both an argument match that starts with `-` and also flag matches that start with `-`.

So the `--` looks like a flag to the parser, and it does not consume the second argument position.

> It works (completes `--` just once and then recognizes the chained command) if I link the `--` arg after each profile arg option as in our current solution.

The second example works because it adds only an argument that starts with `-`.  There are no flags that start with `-`, so `-` is not a "flag character", so the `--` is not interpreted as a flag.



> > But what problem is intended to be solved by calling `--` an argument instead of a flag? Maybe there are other ways to solve the problem?
> 
> I was hoping it would:
> 
> 1. simplify the matcher, see above
> 2. allow autocompleting `--` arg right away (not only after writing the first `-`)
> 
> There's also the question of how to make `--` optional in this case, there doesn't seem to be a dedicated API for this.

`:addarg()` adds an argument position.  There's no way for an argument _position_ to be optional in an argmatcher.

**Reply by elieux** [reply]
Thanks for the clarifications and all the help.

**Comment by chrisant996** [comment]
@elieux Based on the discussion here, I've made several changes to the docs for [Argument Completion](https://chrisant996.github.io/clink/clink.html#argument-completion).  Please take a look and share feedback on the new content.  Thanks!

**Reply by elieux** [reply]
Looks good, thank you.

**Reply by elieux** [reply]
I also noticed that _Functions As Argument Options_ says:

> table of potential matches (**strings**)

... (emphasis mine), while _Delayed initialization for an argument position_ says:

> The table of matches supports the same syntax as `_argmatcher:addarg()`.

I still think the first one is not entirely obvious due to multiple factors, but it's there, so I'm not gonna push it.

**Reply by chrisant996** [reply]
The documentation is different for those because they are very different kinds of functions.

[Functions As Argument Options](https://chrisant996.github.io/clink/clink.html#argmatcher_functions) is a function that is added via `addarg()`, and the function returns matches which get added to a match builder.  It modifies a [match builder](https://chrisant996.github.io/clink/clink.html#builder), so it accepts the same format as [match_builder:addmatches()](https://chrisant996.github.io/clink/clink.html#builder:addmatches).

> **Delayed initialization for an argument position**
> 
> If the overall flags and meaning of the argument positions don't need to be updated, and only the possible values need to be updated within certain argument positions, then you can include `delayinit=function` in the argument table.
> 
> The function should return a table of matches which will be added to the values for the argument position. The table of matches supports the same syntax as [_argmatcher:addarg()](https://chrisant996.github.io/clink/clink.html#_argmatcher:addarg). The function receives two parameters:

The function in `delayinit=function` is not adding matches to a match builder.  It allows an argmatcher to _modify itself_ on the fly, and what it returns is passed into `addarg()` to modify the argmatcher itself.  It modifies an [argmatcher](https://chrisant996.github.io/clink/clink.html#_argmatcher), so it accepts the same format as [argmatcher:addarg()](https://chrisant996.github.io/clink/clink.html#_argmatcher:addarg).

### Function as argument option

```lua
local function list_colors()
    -- This can return the same format that match_builder:addmatches accepts.
    return { "red", "green", "blue" }
end

-- Make argmatcher for "foo", whose first argument provides completions for color names:
clink.argmatcher("foo"):addarg(list_colors)
```

### Delayed initialization for an argument position

```lua
local function list_colors()
    -- This can return the same format that match_builder:addmatches() accepts.
    return { "red", "green", "blue" }
end

-- Function to delay-initialize the first argument position:
local function delay_init()
    -- This can return the same format that argmatcher:addarg() accepts.
    return { list_colors }
end

-- Make argmatcher for "foo", whose first argument position exists but is not initialized yet.
-- The first time the argmatcher is used with a first argument, then the delayinit function is called.
-- When the delayinit function is called, it finishes initializing the argmatcher.
clink.argmatcher("foo"):addarg({delayinit=init_arg})
```
